Validate global V2 templates don't set index.hidden

andreidan opened the pull request.

andreidan committed:
Validate global V2 templates don't set index.hidden

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Indices APIs)

probakowski commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateV2Action.java:
```diff
@@ -88,6 +90,17 @@ public ActionRequestValidationException validate() {
             if (indexTemplate == null) {
                 validationException = addValidationError("an index template is required", validationException);
             }
```
I think we should add return here, otherwise assertion below will trip or we will have NPE if assertions are disabled

andreidan replied:
Ah you're absolutely right, great catch, my brain filled in that return statement. I think an else statement would be better as we'd likely like to fill in all the validation errors

probakowski commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateV2Action.java:
```diff
@@ -88,6 +90,17 @@ public ActionRequestValidationException validate() {
+            if (indexTemplate.indexPatterns().stream().anyMatch(Regex::isMatchAllPattern)) {
+                if (IndexMetadata.INDEX_HIDDEN_SETTING.exists(indexTemplate.template().settings())) {
+                    validationException = addValidationError(
+                        "global V2 templates may not specify the setting " + IndexMetadata.INDEX_HIDDEN_SETTING.getKey(),
```
minor nit: maybe break on `+` here, then there will be 2 lines instead of 4

probakowski commented on server/src/test/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateV2RequestTests.java:
```diff
@@ -44,4 +53,19 @@
         return randomValueOtherThan(instance, this::createTestInstance);
     }
+
+    public void testPutGlobalTemplatesCannotHaveHiddenIndexSetting() {
```
Add test check that `null` template produce correct message and no NPEs

probakowski commented:
Thanks @andreidan for working on this, I've added 2 request around `null` handling and 1 small nit that can be skipped

andreidan committed:
Handle null template properly

dakrone commented:
We could probably also add validation to `MetadataIndexTemplateService.validateTemplate` that can check the current composed_of settings for this also, what do you think @andreidan?

andreidan commented:
@dakrone sure, I'll add that validation in this PR

probakowski reviewed (approved):
LGTM, thanks @andreidan !

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into global-v2template-dont-set-indexhidden

andreidan committed:
Check the index template resolved settings for index.hidden

This also prevents updating the component template to add the index.hidden
setting if that ct is referenced by a global index template.

andreidan commented:
@dakrone made the changes to prevent the `index.hidden` setting from being part of a global template (note that we're preventing it being part of the global template full stop, regardless of the true/false value it might be set to). 

I didn't use the `validateTemplate` method as you initially suggested for this validation as that's more generic, looking at the settings and mappings only, irrespective of what kind of construct it's providing them (index templates v1, v2 or component template). It's also run after the cluster update request was issued but for the purpose of this global indices validation we can do it before we issue the cluster state update (similar to the `validate(PutRequest)` method used by the V1 templates).

andreidan committed:
Add fail statements in test

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into global-v2template-dont-set-indexhidden

andreidan commented:
@dakrone @probakowski this probably slipped off your radar. do you have any more thoughts/suggestions on this?

dakrone reviewed (approved):
LGTM, thanks Andrei!

probakowski commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -196,6 +196,29 @@ ClusterState addComponentTemplate(final ClusterState currentState, final boolean
+            // if the CT is specifying the `index.hidden` setting it cannot be part of any global template
+            if (IndexMetadata.INDEX_HIDDEN_SETTING.exists(finalSettings)) {
+                Map<String, IndexTemplateV2> existingTemplates = currentState.metadata().templatesV2();
+                List<String> globalTemplatesThatUseThisComponent = new ArrayList<>();
+                for (Map.Entry<String, IndexTemplateV2> entry : existingTemplates.entrySet()) {
+                    IndexTemplateV2 templateV2 = entry.getValue();
+                    if (templateV2.composedOf().contains(name) && templateV2.indexPatterns().stream().anyMatch(Regex::isMatchAllPattern)) {
+                        // global templates don't support configuring the `index.hidden` setting so we don't need to resolve the settings as
+                        // no other component template can remove this setting from the resolved settings, so just invalidate this update
+                        globalTemplatesThatUseThisComponent.add(entry.getKey());
+                    }
+                }
```
This can be replaced with stream version, which for me is more readable:
```java
List<String> globalTemplatesThatUseThisComponent = currentState.metadata().templatesV2().entrySet().stream()
                .filter(e -> e.getValue().composedOf().contains(name))
                .filter(e -> e.getValue().indexPatterns().stream().anyMatch(Regex::isMatchAllPattern))
                .map(Map.Entry::getKey)
                .collect(Collectors.toList());
```

andreidan replied:
I'm not sure about this, to me they're equally readable, but your option is a bit more concise. 
In the interest of moving this rather stale PR forward I'm merging it as it is and we can refactor it later if that's ok?

probakowski reviewed (approved):
LGTM, thanks @andreidan! I left few comments but they are purely subjective and can be skipped

andreidan commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -196,6 +196,29 @@ ClusterState addComponentTemplate(final ClusterState currentState, final boolean
+        // if we're updating a component template, let's check if it's part of any V2 template that will yield the CT update invalid
+        if (create == false && finalSettings != null) {
+            // if the CT is specifying the `index.hidden` setting it cannot be part of any global template
+            if (IndexMetadata.INDEX_HIDDEN_SETTING.exists(finalSettings)) {
```
I don't think that'd be necessarily more readable (especially in the context of maybe us wanting to add more checks, this index.hidden one being just one condition we identified now). If it's ok, I'll leave it like this now

probakowski replied:
You definitely can leave it as it is, as I said this totally my personal preference only
